### PR TITLE
101819816 custom tests

### PR DESCRIPTION
**Pull request by vaz**
Total rework of Ra to allow writing custom tests. Based around testing with pytest and WebTest.

Changes:
- A pytest plugin collects test definitions, automatically binding requests to the resource/http method as well as the app.
- Automatic testing can be triggered with `api.autotest()`
- Tests use `example` properties from the RAML by default, but custom object factories can be used to generate request bodies
- Hooks can be declared before/after the suite, tests in a resource, or individual tests, and filtered by resource path or http method.
- Responses are by default automatically validated against the RAML, as well as tested for a 2xx/3xx response code by WebTest (can be overridden)
- Documented in `doc/`

Testing it out (from the README):

Run the test suite:

``` bash
    $ pip install -r requirements.txt
    $ py.test
```

Check out the example:

``` bash
    $ cd examples/ramses-test
    $ pip install -r requirements.txt
    $ py.test
```

The example RAML at `examples/ramses-test/api.raml` and the test file
`examples/ramses-test/tests/test_api.py` should be helpful to reference,
as well as the docs.

### CODE REVIEW

File: ra/dsl.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+        self.hooks = Hooks()
+        self.examples = self._define_factories()
+
+    def _define_factories(self):
+        examples = factory.Examples()
+        for resources in six.itervalues(self.raml.resources):
+            try:
+                post_resource = resources['POST']
+                example = post_resource.body['application/json'].example
```
Comment (postatum) [inline ra/dsl.py]: Will example factory (or other things related to using example) work properly if `post_resource.body['application/json'].example` is `None`?

File: ra/dsl.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        if full_path in self.raml.resources:
+            # get URI param example values:
+            res = list(self.raml.resources[full_path].values())[0]
+            uri_args = raml.uri_args_from_example(res)
+            uri_args.update(uri_params)
+        else:
+            warnings.warn("Declaring resource scope {}: resource not declared "
+                          "in RAML ({})".format(full_path, self.raml_path))
+
+        scope = ResourceScope(path, self,
```
Comment (postatum) [inline ra/dsl.py]: Will `uri_args` be undefined in case `full_path not in self.raml.resources` ?

File: ra/dsl.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                              factory=factory, parent=parent,
+                              **uri_args)
```
Comment (postatum) [inline ra/dsl.py]: Will `uri_args` be not defined if `full_path` not in `self.raml.resources` ?

File: ra/dsl.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+
+    The request object expects to be assigned a ``raml`` attribute with
+    the ra.raml.ResourceNode to validate against as the value.
+
+    :param app:         the app we want to make requests to, generally an
+                        instance of ``webtest.TestApp`` but can be anything
+                        that responds to request() taking a webob-like request
+                        object as a first positional argument, and accepts
+                        request parameters as keyword args.
+    :param base:        the base request class
```
Comment (postatum) [inline ra/dsl.py]: This says "request class" but above `base` is assigned to `ResponseClass`

**Comment by postatum** [comment]
Looks good. I think docs need to describe `exclude` and `only` hook params in more detail.
E.g. when I read docs, I thought `only` is used to specify route paths and `exclude` - HTTP methods.

File: ra/dsl.py (hunk 4 above)
Comment (vaz) [inline ra/dsl.py]: Good catch. Fixed and also simplified to just use webtest's request/response (better methods on response for testing).

File: ra/dsl.py (hunk 3 above)
Comment (vaz) [inline ra/dsl.py]: Yep. Fixed.

**Comment by vaz** [comment]
Thanks for the comments @postatum. Clarified the docs on hooks.

**Comment by postatum** [comment]
When running py.test from example project folder, got:

```
~/projects/ra/examples/ramses-test 
$ py.test
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: --cov=ra --cov-report=term-missing --no-cov-on-fail
  inifile: /home/post/projects/brandicted/ra/tox.ini
  rootdir: /home/post/projects/brandicted/ra
```

Removing pytest.addopts line from tox.ini helped.

Otherwise all looks good.

**Comment by vaz** [comment]
Looks like the pytest-cov plugin is missing in your environment. I'm adding the pytest deps to the requirements file for the example app for completeness' sake (they're already in requirements for ra, so I didn't catch this).

**Comment by jstoiko** [comment]
Good job. The examples/ramses-test no longer works with the current version of ramses. I would delete examples/ anyway because we now have an example in ramses-example (soon to be merged). What do you think?

**Comment by vaz** [comment]
Yes. Deleting examples/, we don't want to have to maintain two almost-identical apps.

**Pull request merged**